### ISSUE TITLE
Out of memory during RPC call

### ISSUE THREAD

**Issue by omani** [begriffs/postgrest#925]
Hi all,

after finishing my project I am experiencing many oom on my server running:
* postgrest
* postgresql
* openresty (nginx reverse proxy)

```
Aug  1 11:56:09 host-test01 kernel: [219140.781331] Out of memory: Kill process 27360 (postgrest) score 415 or sacrifice child
Aug  1 11:56:09 host-test01 kernel: [219140.781582] Killed process 27360 (postgrest) total-vm:1074990692kB, anon-rss:3395440kB, file-rss:0kB
Aug  1 11:56:09 host-test01 kernel: [219140.832559] systemd invoked oom-killer: gfp_mask=0x201da, order=0, oom_score_adj=0
Aug  1 11:56:09 host-test01 kernel: [219140.832563] systemd cpuset=/ mems_allowed=0
```

this is part of the kernel call trace. it says that postgrest will be sacrified to get free memory.

What am I doing:
* I am writing files to and from postgresql (in a table) chunk-wise
* mostly rpc calls (read_blob, write_blob)

hardware specs:
* 2 vCPUs (cores)
* 8GB RAM
* postgrest, openresty and postgresql in docker containers

any hints on what we can do here? where to start debugging and what to do to solve this problem?

starting the postgrest docker in attached mode shows the usual printing log lines (http requests) and as soon as the kernel kills the container I just drop out the container shell and the container stops running.

**Comment by omani** [comment]
if you need more information please tell me and I will provide if I can.

**Comment by ruslantalpa** [comment]
this seems to be a problem with how the binary was build (built on ubuntu box, used in debian docker).
please reopen if a proper build does not solve this

**Comment by omani** [comment]
after building postgrest for ubuntu:16.04 as described here: https://postgrest.com/en/v4.1/install.html#build-from-source

I am still suffering from out of memory issues.

**Comment by ruslantalpa** [comment]
so you are building the binary on a ubuntu box and then you copy that binary to a docker image that is based on the same ubuntu?

**Comment by omani** [comment]
yes

**Comment by omani** [comment]
please see https://nopaste.linux-dev.org/?1160385 for my prof file.

**Comment by omani** [comment]
the line is: ```trimNullChars = T.takeWhile (/= '\x0')```

**Comment by omani** [comment]
switched from rpc to view with a trigger. solved.

**Comment by begriffs** [comment]
Hmm, glad you found another way to do it, but wondering if this is still an RPC problem that other people will run into.

**Comment by omani** [comment]
I am sure people will run into this problem when using rpcs. @9too, @ruslantalpa and me discussed this in gitter. maybe you guys can get together and spend time on this. for me this is solved now but I think it is worth to investigate nevertheless.

**Comment by begriffs** [comment]
OK, makes sense. Well I'm going to open this issue again since it has useful information for the ongoing RPC problem.

**Comment by ruslantalpa** [comment]
Looking at https://hackage.haskell.org/package/interpolatedstring-perl6-1.0.0/docs/src/Text-InterpolatedString-Perl6.html
there is a lot of recursions and reversing going on, probably where the high memory usage comes from.

While these symptoms are avoided for insert, they can not be avoided for rpc and update.
we use `qc` mainly for "nice code" but i don't think it was designed to extremely long strings.

As a first step, we should probably remove `qc` from the code and see how it goes.

@omani plese keep your old rpc code to be able to run a custom branch to see if there is progress when removing the lib and going with simple string concatenation.

**Comment by 9too** [comment]
Here's a minimal working example : 
```sql
CREATE FUNCTION api.leak(blob bytea) RETURNS void
AS $$ BEGIN END; $$ LANGUAGE plpgsql;
```
```sh
( echo 'blob=>'; base64 postgrest; ) > input
while true; do curl -d @- 'http://localhost:3000/rpc/leak' < input; done;
```

Basically, you really need **big** chunks.

**Comment by begriffs** [comment]
Thanks for the steps for reproducing. I wonder if this has a similar cause as #690

**Comment by steve-chavez** [comment]
I think this is more related to `trimNullChars` than `qc`, the custom input sanitizing can be avoided with a parametrized query;  using [this function](https://github.com/begriffs/postgrest/blob/master/test/fixtures/schema.sql#L1292-L1294) as an example, I've come up with this new query for `callProc`:
```sql
with args as (
  select * from json_to_record('{"num":1, "str": "smth", "b": true}') as _(num int, str text, b boolean)
),
func as (
  select test.many_inout_params(num := (select num from args), str := (select str from args), b:= (select b from args))
)
select * from func;
```
The key is the `json_to_record` function(avalaible from pg >= 9.4), with this we could input the args as a `$1`. I'll change the `callProc` query to this one and then compare profiling reports.

**Comment by ruslantalpa** [comment]
we already use json_populate_recordset https://github.com/begriffs/postgrest/blob/master/src/PostgREST/QueryBuilder.hs#L290

can you use that?

PS: interesting approach but somehow is seems too elaborate :) though i don't mind it if it works out

**Comment by steve-chavez** [comment]
The problem is that `json_populate_recordset` needs a type in the first arg like `null::type` and I didn't found a way to specify a "dynamic type" in that way so `json_to_record` is convenient because it can specify that.

**Comment by steve-chavez** [comment]
New `callProc` query solves the problem, at least now I'm getting practically the same memory usage as when using INSERT. I'll show some profile reports proving this, I'll be using the function provided [here](https://github.com/begriffs/postgrest/issues/925#issuecomment-320942107) and for contrasting with INSERT I'm using this table:
```sql
create table leak(
  id serial primary key,
  blob bytea
);
```
With this `100mb` file as payload:
```bash
dd if=/dev/zero bs=74M count=1 | base64 > 100mb.bin
```
I'm getting these profile reports on an Archlinux x64 - i7 - 8GB RAM:
- Doing a `POST rpc/leak` with old callProc query, using:
```bash
echo "{\"blob\": \"$(cat 100mb.bin)\"}" | curl -H "Content-Type: application/json" -d @ http://localhost:3000/rpc/leak
```  
```
     postgrest +RTS -N -I2 -p -h -RTS config

  total time  =        6.62 secs   (6616 ticks @ 1000 us, 1 processor)
  total alloc = 22,266,823,688 bytes  (excludes profiling overheads)

COST CENTRE     MODULE                            SRC                                                     %time %alloc

trimNullChars   PostgREST.QueryBuilder            src/PostgREST/QueryBuilder.hs:478:1-38                   33.2   48.3
toQQ            Text.InterpolatedString.Perl6     src/Text/InterpolatedString/Perl6.hs:200:5-29            33.1   36.5
>>=.\.succ'     Data.Attoparsec.Internal.Types    Data/Attoparsec/Internal/Types.hs:146:13-76              10.0    1.4
replace.ixs     Data.Text                         Data/Text.hs:691:5-33                                     7.3    0.0
append.x        Data.Text                         Data/Text.hs:(464,7)-(468,18)                             7.3   10.2
hashPtrWithSalt Data.Hashable.Class               Data/Hashable/Class.hs:(714,1)-(716,23)                   3.2    0.0
pgFmtLit        PostgREST.QueryBuilder            src/PostgREST/QueryBuilder.hs:(197,1)-(203,15)            3.2    0.0
append.\.\      Data.Attoparsec.ByteString.Buffer Data/Attoparsec/ByteString/Buffer.hs:(110,35)-(134,56)    0.2    1.2
encodeUtf8      Data.Text.Encoding                Data/Text/Encoding.hs:(373,1)-(388,37)                    0.0    1.9
```
- Doing an INSERT to the leak table `POST /leak` with:
```bash
echo "{\"blob\": \"$(cat 100mb.bin)\"}" | curl -H "Content-Type: application/json" -d @ http://localhost:3000/leak
```
```
     postgrest +RTS -N -I2 -p -h -RTS config

  total time  =        2.51 secs   (2506 ticks @ 1000 us, 1 processor)
  total alloc = 5,872,068,952 bytes  (excludes profiling overheads)

COST CENTRE           MODULE                            SRC                                                     %time %alloc

unquoted              Data.Aeson.Encoding.Builder       Data/Aeson/Encoding/Builder.hs:119:1-47                  67.4   84.6
>>=.\.succ'           Data.Attoparsec.Internal.Types    Data/Attoparsec/Internal/Types.hs:146:13-76              25.7    5.3
receiveloop           Network.Wai.Handler.Warp.Recv     Network/Wai/Handler/Warp/Recv.hs:(115,1)-(129,20)         1.7    0.0
append.\.\.\          Data.Attoparsec.ByteString.Buffer Data/Attoparsec/ByteString/Buffer.hs:(127,40)-(134,56)    1.6    0.0
append.\.\            Data.Attoparsec.ByteString.Buffer Data/Attoparsec/ByteString/Buffer.hs:(110,35)-(134,56)    0.9    4.6
sendQueryPrepared.\.\ Database.PostgreSQL.LibPQ         src/Database/PostgreSQL/LibPQ.hsc:(1705,16)-(1709,69)     0.9    1.8
encode                Data.Aeson                        Data/Aeson.hs:128:1-46                                    0.0    1.8
nullableValue.\.\     Hasql.Private.Encoders.Params     library/Hasql/Private/Encoders/Params.hs:51:37-78         0.0    1.8
```
- With the new callProc query, it's on this [branch](https://github.com/steve-chavez/postgrest/tree/rpc-memory)(dirty and leak function hardcoded for now), using the same bash command as before to `POST rpc/leak` we get:
```
           postgrest +RTS -N -I2 -p -h -RTS config

        total time  =        2.49 secs   (2490 ticks @ 1000 us, 1 processor)
        total alloc = 5,871,934,768 bytes  (excludes profiling overheads)

      COST CENTRE           MODULE                            SRC                                                     %time %alloc

      unquoted              Data.Aeson.Encoding.Builder       Data/Aeson/Encoding/Builder.hs:119:1-47                  68.5   84.6
      >>=.\.succ'           Data.Attoparsec.Internal.Types    Data/Attoparsec/Internal/Types.hs:146:13-76              24.4    5.3
      receiveloop           Network.Wai.Handler.Warp.Recv     Network/Wai/Handler/Warp/Recv.hs:(115,1)-(129,20)         1.8    0.0
      append.\.\.\          Data.Attoparsec.ByteString.Buffer Data/Attoparsec/ByteString/Buffer.hs:(127,40)-(134,56)    1.8    0.0
      sendQueryPrepared.\.\ Database.PostgreSQL.LibPQ         src/Database/PostgreSQL/LibPQ.hsc:(1705,16)-(1709,69)     0.8    1.8
      append.\.\            Data.Attoparsec.ByteString.Buffer Data/Attoparsec/ByteString/Buffer.hs:(110,35)-(134,56)    0.5    4.6
      encode                Data.Aeson                        Data/Aeson.hs:128:1-46                                    0.1    1.8
      nullableValue.\.\     Hasql.Private.Encoders.Params     library/Hasql/Private/Encoders/Params.hs:51:37-78         0.0    1.8
``` 
I've used `json` instead of `application/x-www-form-urlencoded` for the tests because with the latter almost 3x of the memory is used for both INSERT and RPC, that's because of the querystring parsing, so I've removed that extra work because really it isn't relevant to fixing the callProc query.

**Comment by steve-chavez** [comment]
So I'll change the code to use the new callProc query to avoid the custom input sanitizing and the high memory usage.

Also wanted to comment that right now the memory usage is really high for big payloads, for a 100mb file we get like 6GB in memory allocation, also I've tried uploading a 300MB file and couldn't get the operation to finish, it used all my available ram plus swap and leaved my system unresponsive so I had to kill postgrest.

To avoid the Aeson json encoding, that does most of the allocating, maybe we could come up with an interface to upload raw binary, this was requested in #922 and #894.
